otherlibs/dune-rpc/dune_rpc.mli

```diff
@@ -244,6 +250,22 @@ module V1 : sig
 
       val notification : t -> 'a Notification.t -> 'a -> unit fiber
 
+      module Subscription : sig
+        type t
+
+        val await : t -> unit fiber
+
+        val cancel : t -> unit fiber
+      end
+
+      val subscribe :
+           ?id:Id.t
+        -> t
+        -> ('a, 'b) Sub.t
+        -> on_init:('a -> Subscription.t -> 'res fiber)
```

Comment:
This feels a bit convoluted. Is the intention that `'res` is to be an accumulator, so it acts as if we're folding over the stream? Does the `unit fiber` returned at the end resolve when the subscription is ack'd, or when the subscription is cancelled (from the server end? by the client?)

Reply:
I should definitely comment this. My intention with `'res` is to guarantee that `on_init` must terminate before any `on_next` are called. It can be done without the extra return value from `on_init`, but I think it's nice that the types help demonstrate this invariant.

The final `unit fiber` is returned when the subscription is terminated. This can happen in two ways:
* The client cancels with `Subscription.cancel`.
* The server notifies the client no more values will be sent.

As for folds, I've thought about implementing this interface:
```ocaml
val subscribe :  .. on_init:('a -> Subscrition -> 'res fiber) -> on_next:('a -> 'res -> 'res fiber) -> 'res fiber
```

But to me this seems a bit limiting. Why should we wait for an `on_next` to finish before calling `on_next` for the following value? We can implement something similar to `Fiber.map_reduce_errors` to reflect the semantics we want better, but it seems a bit heavy?

Perhaps @aalekseyev can glance at this function and give a suggestion?

Reply:
So, if I understand this right, `on_init` is intended to initialize some (mutable, by the sounds of it?) state of type `'res`, which is then passed to `on_next`, and invocations of `on_next` are independent?

This feels like it's asking for race conditions if we don't confirm that a previous `on_next` has completed.

Reply:
> So, if I understand this right, on_init is intended to initialize some (mutable, by the sounds of it?) state of type 'res, which is then passed to on_next, and invocations of on_next are independent?

No mutability is needed. If I implied it somehow, it was a mistake. `'res` is any (immutable or mutable) value that is passed to `on_next` to acknowledge that `on_init` is finished.

The guarantee for `on_next` is that it is called in sequence. It does not necessarily mean that they complete in sequence. Although you could implement this property yourself with `Fiber.Mutex.t` for example:
```ocaml
let on_init _ _ = Fiber.Mutex.create ()
let on_next a mutex = Fiber.Mutex.with_lock mutex (fun () -> f a)
```
But note that this will create unbounded queuing. In particular, you'll eventually run out of memory if your `on_next` is processing slower than you're receiving `a`. With the current api, your `on_next` is guaranteed to immediately fire (but in order)

Reply:
> No mutability is needed. If I implied it somehow, it was a mistake.

Sorry, in the `unit` variant, that suggests that there's some kind of effect happening, no? Otherwise, I can't see how `'a -> 'res -> unit fiber` is a useful type for the handler.

Reply:
Okay, I see what you mean. There are plenty of effects available in `unit fiber` other than mutability. But as I mentioned earlier, we could do the map reduce based API that allows to process the stream without unbounded queuing.

Reply:
I think it's better if the API is sequential by default with parallelism "opt in" than the opposite way around. At least with the Pipe RPC and State RPC we're consuming pipes in order and it's very rare to need parallelism. When we do, we can always "spawn" an asynchronous computation and yield control so the next element of the pipe is processed concurrently.

Also, I think there's value in the intermediate type being returned (Pipe.t in the case of Async rpc), so that the call site can reason in terms of a first-class stream instead of thinking about callbacks. (although the difference is not big and may be a matter of perspective).

Reply:
As for the unbounded queuing, I don't see that as something we're solving by introducing a parallel API. Why is queuing jobs in the scheduler any better than queuing elements in the pipe?

Reply:
I tried implementing the sequential API, and I'm not sure how to sequentialize the operations without asking the user to provide an `Mvar` implementation. 

> Also, I think there's value in the intermediate type being returned (Pipe.t in the case of Async rpc), so that the call site can reason in terms of a first-class stream instead of thinking about callbacks. (although the difference is not big and may be a matter of perspective).

I prefer that as well, but this API need to be very portable and as a consequence we have no appropriate stream abstraction that we can use. We could of course ask the user to provide a pipe/stream signature, but they can just as easily use the callback api to create the pip/stream.

@cwong-ocaml do you have a preference for a particular api?

Reply:
> I'm not sure how to sequentialize the operations without asking the user to provide an Mvar implementation

I can't follow this argument, but if Mvar makes things better then it seems fine to ask for it? After all, that's a simpler primitive than a stream.

Reply:
I'm also not sure I follow the Mvar conversation.

From a purely selfish perspective, I would prefer to provide a more featured implementation that the client can then take full advantage of (e.g., providing a stream implementation so subscriptions can produce a stream). However, I would also generally prefer to force fewer abstractions upon users if it can be avoided.

Reply:
> I can't follow this argument, but if Mvar makes things better then it seems fine to ask for it? After all, that's a simpler primitive than a stream.

It's simple: I do not know how to implement the semantics you desire without an Mvar. Now I have no problem with asking the user to provide an Mvar module to the functor, but I just dislike that the user has to do so even if they aren't planning to use the subscription mechanism.

Reply:
> From a purely selfish perspective, I would prefer to require a more featured implementation that the client can then take full advantage of (e.g., providing a stream implementation so subscriptions can produce a stream). However, I would also generally prefer to force fewer abstractions upon users if it can be avoided.

Wouldn't you prefer to use your own stream implementation rather than use the one provided by this tiny library? In the async world, it is my understanding that `Pipe.t` is commonly used as the stream piece of plumbing. I assumed you would have preferred to use that.

Reply:
Oh, that's what I mean -- I would say `'a stream = 'a Pipe.t`, and then any subscription functions would then give me a `Pipe.t`. I'm aware that they aren't strictly the same abstraction, but it's more convenient to put the conversion in the `Fiber` input rather than as a wrapper around the RPC in general (as we're currently doing).

Reply:
> It's simple: I do not know how to implement the semantics you desire without an Mvar.

I mean, I got that, but I don't know the details of your idea how you'd use Mvar.

Reply:
I pushed a commit to demonstrated what I intended. Now that I think about it, I could replace `Mvar` with `Stream` and change the interface to:

```ocaml
module Stream : sig
  type 'a t
  val create : unit -> 'a t
  val write : 'a t -> 'a -> Fiber.t
  val close : 'a t -> unit Fiber.t
end

    val subscribe :
         ?id:Id.t
      -> t
      -> ('a, 'b) Sub.t
      -> ('a * Subscription.t * 'b Stream.t) fiber
```

Reply:
> Now that I think about it, I could replace Mvar with Stream and change the interface to:

I quite like this version.

Reply:
I pushed a commit that implements this. Happy to go along with this interface if that's what is preferred @cwong-ocaml @aalekseyev